From 086bc5143a8377379c9e8f07230ca57639b8425c Mon Sep 17 00:00:00 2001
From: Danny Baumann <dannybaumann@web.de>
Date: Wed, 15 Aug 2018 12:10:34 +0200
Subject: [PATCH 1/2] Don't request battery optimization exemption if not
 needed.

Commit I585b791ff44c5dbc4b2503b9bcc1a132d18d4856 added requesting
battery optimization exemption if IMAP IDLE is used. It missed the fact
that the IDLE restart intent is also sent (on connection changes) if no
account is configured for IMAP IDLE. Fix it.

Change-Id: Icdac9eebc030b56df17a1774fa5bccffae5a53b4
---
 .../com/android/email/service/ImapService.java        | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/provider_src/com/android/email/service/ImapService.java b/provider_src/com/android/email/service/ImapService.java
index 02c35e94b..22f449790 100644
--- a/provider_src/com/android/email/service/ImapService.java
+++ b/provider_src/com/android/email/service/ImapService.java
@@ -790,16 +790,13 @@ public class ImapService extends Service {
             // been changes while we lost connectivity. At the end of the sync
             // the IDLE connection will be re-established.
 
-            if (!checkBatteryOptimizationsExemption(this)) {
-                return START_STICKY;
-            }
-
             mIdleRefreshWakeLock.acquire();
 
             sExecutor.execute(new Runnable() {
                 @Override
                 public void run() {
                     ContentResolver cr = context.getContentResolver();
+                    List<Account> syncAccounts = new ArrayList<>();
                     Cursor c = null;
                     try {
                         c = cr.query(Account.CONTENT_URI, Account.CONTENT_PROJECTION,
@@ -814,6 +811,12 @@ public class ImapService extends Service {
                             // Only imap push accounts
                             if (account.getSyncInterval() == Account.CHECK_INTERVAL_PUSH
                                     && isLegacyImapProtocol(context, account)) {
+                                syncAccounts.add(account);
+                            }
+                        }
+                        if (!syncAccounts.isEmpty()
+                                && checkBatteryOptimizationsExemption(context)) {
+                            for (Account account : syncAccounts) {
                                 requestSyncForAccountMailboxesIfNotIdled(account);
                             }
                         }
-- 
2.17.1

